genie-common-internal/src/main/java/com/netflix/genie/common/internal/util/FileBufferFactory.java

```diff
@@ -0,0 +1,284 @@
+
+/**
+ * Factory of {@link FileBuffer} which provides its own implementation of the interface.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class FileBufferFactory {
+
+    private static final String TEMP_FILE_PREFIX = FileBuffer.class.getSimpleName() + "-";
+    private static final String TEMP_FILE_SUFFIX = null;
+
+    /**
+     * Get a new {@link FileBuffer}.
+     *
+     * @param size the exact size the file buffer is expected to eventually reach
+     * @return a new {@link FileBuffer}
+     * @throws IOException in case of error creating the file buffer
+     */
+    public FileBuffer get(final int size) throws IOException {
+        final File tempFile = Files.createTempFile(TEMP_FILE_PREFIX, TEMP_FILE_SUFFIX).toFile();
+        tempFile.deleteOnExit();
+        return new FileBufferImpl(tempFile, size);
+    }
+
+    private static class FileBufferImpl implements FileBuffer {
+        private final File tempFile;
+        private final InputStream inputStream;
+        private final OutputStream outputStream;
+
+        FileBufferImpl(final File tempFile, final int expectedFileSize) throws FileNotFoundException {
+            this.tempFile = tempFile;
+            this.inputStream = new FileBufferInputStream(tempFile, expectedFileSize);
+            this.outputStream = new FileBufferOutputStream(tempFile, expectedFileSize);
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public InputStream getInputStream() {
+            return this.inputStream;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public OutputStream getOutputStream() {
+            return this.outputStream;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public File getFile() {
+            return this.tempFile;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void close() throws IOException {
+            this.inputStream.close();
+            this.outputStream.close();
+            final boolean deleted = this.tempFile.delete();
+            if (!deleted) {
+                log.warn("File not deleted: {}", this.tempFile.getAbsolutePath());
+            }
+        }
+
+        /**
+         * Implementation of InputStream that relies on the pre-determined file size.
+         * - Blocks read until more data is available
+         * - Returns -1 once the entire file content has been read
+         */
+        private static class FileBufferInputStream extends InputStream {
+            private final FileInputStream fileInputStream;
+            private final int expectedFileSize;
+            private AtomicBoolean closed = new AtomicBoolean(false);
+            private AtomicInteger watermark = new AtomicInteger(0);
+
+            FileBufferInputStream(final File tempFile, final int expectedFileSize) throws FileNotFoundException {
```

Comment:
What I'm passing around is a temporary file, it just got created and later it will be deleted.
I can always go from path back to file, but logically, it's a file I'm passing around.

I'm going to attempt to switch from FileBuffers to Channels so it's using the newer APIs.
Tho using the asynchronous operations in this case would just make things more complicated and i don't see much benefit.

Reply:
Stopped using `File` and `FileWriter` in favor of `nio` APIs.
Replacing streams with channels is a bigger change that I'm not going to do for the moment (mostly because we cannot go all-in: the API we need to expose to Spring still uses `InputStream`).